### PR TITLE
chore: add more validation to `MsgRun` & `MsgAddPackage`

### PR DESCRIPTION
**Pull request by leohhhn**
## Description

This PR adds more validation to `MsgRun` & `MsgAddPkg`. It does this by adding a few checks to `MemPackage`'s `Validate()`, and then incorporating checks on `MemPackage` within `MsgRun` & `MsgAddPkg`.

<details><summary>Contributors' checklist...</summary>

- [x] Added new tests, or not needed, or not feasible
- [x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [x] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

File: tm2/pkg/std/memfile.go [hunk 1]
```diff
@@ -50,12 +50,19 @@ var (
...
+	}
 	if !rePkgName.MatchString(mempkg.Name) {
 		return errors.New(fmt.Sprintf("invalid package name %q", mempkg.Name))
 	}
 	if !rePkgOrRlmPath.MatchString(mempkg.Path) {
 		return errors.New(fmt.Sprintf("invalid package/realm path %q", mempkg.Path))
 	}
+	if len(mempkg.Files) == 0 {
+		return errors.New(fmt.Sprintf("package contains no files"))
+	}
```
Comment (harry-hov) [inline tm2/pkg/std/memfile.go:64]: maybe?
```suggestion
	if memPkg.IsEmpty() {
		return errors.New(fmt.Sprintf("found an empty package %q", mempkg.Path))
	}
```

Comment (leohhhn) [reply tm2/pkg/std/memfile.go]: [dc6bf24](https://github.com/gnolang/gno/pull/1646/commits/dc6bf24ef8c5daffb3e22b592516059b090234b7)

**Comment by zivkovicmilos** [comment]
Please give me a minute to review this, I'm sensing it's a touchy PR 🙏

**Review by gfanton — APPROVED** [review]

**Comment by thehowl** [comment]
cc/ @jaekwon, from review meeting:

Is there any reason why `ValidateBasic` is called basic? Should the checks perform within it be very fast, "basic" or whatever? Can we add the `mempkg` check in there, or would it create problems?

File: tm2/pkg/std/memfile.go [hunk 2]
```diff
@@ -50,12 +50,19 @@ var (
 // file names must contain dots.
 // NOTE: this is to prevent conflicts with nested paths.
 func (mempkg *MemPackage) Validate() error {
+	if mempkg == nil {
+		return errors.New(fmt.Sprintf("package is nil"))
```
Comment (zivkovicmilos) [inline tm2/pkg/std/memfile.go:54]: You should format errors with `fmt.Errorf`.

Looking at this again, you don't even format an error here, this entire call is useless - define it as a package-level error

File: tm2/pkg/std/memfile.go [hunk 3]
```diff
@@ -50,12 +50,19 @@ var (
...
+	}
 	if !rePkgName.MatchString(mempkg.Name) {
 		return errors.New(fmt.Sprintf("invalid package name %q, failed to match %q", mempkg.Name, rePkgName))
 	}
 	if !rePkgOrRlmPath.MatchString(mempkg.Path) {
 		return errors.New(fmt.Sprintf("invalid package/realm path %q, failed to match %q", mempkg.Path, rePkgOrRlmPath))
 	}
+	if mempkg.IsEmpty() {
+		return errors.New(fmt.Sprintf("package %q contains no files", mempkg.Name))
```
Comment (zivkovicmilos) [inline tm2/pkg/std/memfile.go:63]: You should format errors with `fmt.Errorf`

File: tm2/pkg/std/memfile.go [hunk 4]
```diff
@@ -50,12 +50,19 @@ var (
 // file names must contain dots.
 // NOTE: this is to prevent conflicts with nested paths.
 func (mempkg *MemPackage) Validate() error {
```
Comment (zivkovicmilos) [inline tm2/pkg/std/memfile.go:52]: Please add unit tests for `MemPackage` validation

File: gno.land/pkg/sdk/vm/msgs.go [hunk 5]
```diff
@@ -59,6 +59,10 @@ func (msg MsgAddPackage) ValidateBasic() error {
 	if !msg.Deposit.IsValid() {
 		return std.ErrTxDecode("invalid deposit")
 	}
+	if err := msg.Package.Validate(); err != nil {
+		return ErrInvalidPackage(err.Error())
+	}
```
Comment (zivkovicmilos) [inline gno.land/pkg/sdk/vm/msgs.go:64]: Where is this covered?

File: gno.land/pkg/sdk/vm/msgs.go [hunk 6]
```diff
@@ -187,6 +191,10 @@ func (msg MsgRun) ValidateBasic() error {
 		return ErrInvalidPkgPath(fmt.Sprintf("invalid pkgpath for MsgRun: %q", path))
 	}
 
+	if err := msg.Package.Validate(); err != nil {
+		return ErrInvalidPackage(err.Error())
+	}
```
Comment (zivkovicmilos) [inline gno.land/pkg/sdk/vm/msgs.go:196]: Where is this covered?

File: gno.land/pkg/sdk/vm/errors.go [hunk 7]
```diff
@@ -13,16 +13,22 @@ type (
 	InvalidPkgPathError struct{ abciError }
 	InvalidStmtError    struct{ abciError }
 	InvalidExprError    struct{ abciError }
+	InvalidPackageError struct{ abciError }
 )
 
 func (e InvalidPkgPathError) Error() string { return "invalid package path" }
+func (e InvalidPackageError) Error() string { return "package validation failed" }
```
Comment (zivkovicmilos) [inline gno.land/pkg/sdk/vm/errors.go:20]: Please add a documentation reference in this PR as to what is a "valid package" 🙏

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
Why would you move the `Package` validation outside the keeper methods, that validate it upon execution, and into `ValidateBasic`, which should be a quick sanity check for validity?

We use `Package.Validate` and `ValidateBasic` in the same sense how we use `CheckTx` and `DeliverTx` - one is a quick sanity check, and the other is detailed validation (VM processing).

We execute `ValidateBasic` for top-level validations (ex. when the node receives a tx message from gossip), and do additional validations (execution validations) when the node actually commits it to the VM.

I am not convinced this change is valid

**Comment by zivkovicmilos** [comment]
@leohhhn what's the status on this PR?

cc @Kouteki

**Comment by leohhhn** [comment]
@zivkovicmilos 

This has been hanging mainly because I assume I will spend too much time due to my lack of knowledge on how the keeper works - I will give it a shot this week and try to incorporate the change you requested. Will ping you when I'm done 👍

**Pull request closed without merging**